rubicon/objc/api.py

```diff
@@ -968,6 +1000,24 @@ def _new_from_class_statement(cls, name, bases, attrs, *, protocols):
             else:
                 class_register(ptr, attr_name)
 
+        # Add cleanup of ivars / properties to dealloc
+
+        old_dealloc = libobjc.class_getMethodImplementation(ptr, SEL("dealloc"))
+
+        def _new_delloc(objc_self, _cmd):
+
+            # Invoke dealloc callback of each property.
+            for attr_name, obj in attrs.items():
+                if isinstance(obj, objc_property):
+                    obj.dealloc_callback(objc_self, attr_name)
+
+            # Invoke original dealloc.
+            cfunctype = CFUNCTYPE(None, objc_id, SEL)
+            old_dealloc_callable = cast(old_dealloc, cfunctype)
+            old_dealloc_callable(objc_self, SEL("dealloc"))
+
+        add_method(ptr, "dealloc", _new_delloc, [None, ObjCInstance, SEL], replace=True)
```

Comment:
What behavior do we want if a class contains both a user-defined `dealloc` method and properties with dealloc callbacks? The implementation here will call the dealloc callbacks before the user-defined `dealloc` method. But in practice I think the opposite order would be more useful - that way the user-defined `dealloc` could call methods on objects stored in properties before they are released by the callbacks (which could cause them to be deallocated right away).

Implementing this would be a bit more difficult. A user-defined `dealloc` is expected to end with a call to the superclass `dealloc`. But if a class has dealloc callbacks, those would need before the superclass `dealloc` (but also after the user-defined `dealloc` code). There's no way for Rubicon to insert the dealloc callbacks into the user's `dealloc` method, so the user would have to manually call the dealloc callbacks before the super call (via an extra function/method provided by Rubicon).

A different solution would be to allow users to define their own dealloc callback that gets called before any dealloc callbacks from properties. Then users could use that callback instead of manually overriding `dealloc`, and let Rubicon generate a `dealloc` method that calls all callbacks in the right order.

Reply:
I agree that calling the user's `dealloc` before clearing any instance variables can make life at lot easier for the user. But I don't really like either of the proposed solutions, both look like workarounds that force the user to become aware of the memory management which we are otherwise performing behind the scenes.

I can possibly think of two alternative approaches that could provide a better user experience (but would complicate our own implementation):

1. We document that user should not call the `dealloc` of the super class in their own implementation, as in a proper ARC environment. We then call the following methods in order: First, any user-defined dealloc, second, our cleanup code and third the super class's dealloc. This would mean treating the dealloc definition differently from other method definitions, at least internally. It would however be transparent to the user.

2. We call our own cleanup code *after* calling the old `delloc` implementation (which includes the user's code and any calls to the super dealloc). I'm not if this is possible to do reliably.

What do you think, is either of those options worth the effort?

Reply:
Option 1 would be ideal, because it's simple to use and matches what you do in normal ARC Objective-C code. But implementing this behavior now would silently break existing code that overrides `dealloc` and correctly calls the superclass `dealloc` at the end - that would cause the superclass `dealloc` to get called twice, which can lead to other objects being `release`d too often and other similar problems. That's the main reason why I suggested adding a separate user-definable callback for the same purpose.

Rubicon is still before version 1.0, so we *could* still make breaking changes like this, especially if it improves usability in the long term. But if we do that, we should try to throw errors for code relying on the old behavior, to avoid silent double frees in code that was previously correct.

I honestly don't know if option 2 would be safe or not... There *should* be no way for the superclass `dealloc`s to corrupt ivars defined by the subclass, but I wouldn't really rely on it. Especially `NSObject`'s top-level `dealloc` could do things that make the entire object unusable somehow.

Another alternative would be to not touch user-defined `dealloc`s at all, and only generate an automatic `dealloc` if the user hasn't already overridden `dealloc` manually. This would be fully compatible with existing code and should also be simple to implement. The disadvantage is that if the user really needs to add custom code to `dealloc`, they then have to manually do all the cleanup that the dealloc callbacks would have done automatically.

Reply:
I do prefer Option 1 from those choices, or not touching use-defined deallocs at all together with a clear documentation on manual cleanup. My only issue with Option 1 is, how can we raise an error if user does call the superclass dealloc? Is there an elegant way of doing so? We are not a compiler after all and don't want to inspect the actual code in the user's dealloc.

Reply:
I can confirm that running our cleanup *after* the superclass dealloc does lead to trouble during the ivar cleanup. In particular, `libobjc.class_getInstanceVariable()` returns `None`. So this is not viable.

I've implemented the first option now: calling the user's dealloc, then our cleanup, then the superclass dealloc. For the time being, there is no special error handling if the user calls the superclass dealloc manually. It will however raise (obscure) errors when we run our own cleanup. Without our own cleanup code (for example when no properties are declared), dealloc is called twice and leads to a segfault. In either case, users will notice that *something* is wrong without knowing what it might be. Not ideal...

Reply:
The best workaround I can come up with, short of actually inspecting the user's dealloc code when creating the class, is to print a warning when `send_super` is called, before either segfaulting or failing to complete the dealloc. What do you think, is this acceptable?

Reply:
Sorry for the late reply - had some uni stuff that I needed to finish in the last few days.

The implementation you wrote looks good IMO. You're right that we can't do much to guard against old code that still calls `send_super` at the end of `dealloc`. The only option I can think of is what you've already suggested - add a special case inside `send_super` that warns whenever `send_super(self, "dealloc", ...)` is called. If we do that, we just need to make sure that the warning doesn't appear when Rubicon itself calls `send_super(self, "dealloc", ...)` - probably using an internal keyword to suppress the warning?

That way I think we could even "fix" the segfaults. If `send_super(self, "dealloc", ...)` is called *without* the special internal keyword argument, we can make it show a warning and then return without actually calling the super method. That way, if a user-defined `dealloc` calls `send_super`, the super `dealloc` won't actually be called yet. Then the user-defined `dealloc` returns to Rubicon, which runs the dealloc callbacks then makes the real `send_super` call (with the internal keyword argument, so that this time it actually calls the super method).

This isn't a very nice solution - but if it works, I would rather have some less nice code in Rubicon that shows a helpful warning about the change, rather than breaking existing code so that it causes unexplained segfaults.